docs/developers/controller-logic.md

```diff
@@ -0,0 +1,69 @@
+## Kubernetes objects
+
+Before learning about how Tekton works, it's useful to take some time to understand what a Kubernetes object is.
+Please see [Understanding Kubernetes objects](https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/)
+for an overview of working with objects.
+Kubernetes [API conventions docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds)
+are another useful resource for understanding object terminology.
+
+## Tekton CRDs, Webhooks, and Controllers
+
+### CRDs
+
+Tekton objects, including Tasks, TaskRuns, Pipelines, PipelineRuns, and other resources, are implemented as
+Kubernetes [CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/),
+which are extensions of the Kubernetes API.
+To better understand what it means to create a CustomResourceDefinition (CRD), check out the tutorial
+[Extend the Kubernetes API with CustomResourceDefinitions](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/).
+Tekton CRDs are defined in the [config folder](../../config/), and the schemas of these CRDs are defined as Go structs in the [apis folder](../../pkg/apis/pipeline/v1).
+
+### Controllers
+
+Creating an instance of a CRD doesn't "do" much. As shown in the [CRD tutorial](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/),
+creating a CRD allows you to list Objects of that Kind but doesn't result in anything executing.
+As described in ["custom controllers"](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#custom-controllers),
+you must implement a controller to change a Kubernetes cluster's state when an instance of a CRD is created, a process called "reconciling".
+Reconcilers change the cluster based on the desired behavior defined in an object's "spec", and update the object's "status" to reflect what happened.
+
+The [Kubebuilder tutorial](https://book.kubebuilder.io/introduction.html) is a helpful resource for understanding what it means to build a controller
+that reconciles CRDs. (Note: Tekton does not use Kubebuilder.)
+
+Not all Tekton CRDs have controllers. For example, there is no reconciler for Tasks, meaning that creating a Task doesn't "do" anything.
+To run a Task, you must create an instance of a TaskRun referencing it, and the TaskRun is executed by the TaskRun reconciler.
+
+The TaskRun reconciler code is found in pkg/reconciler/taskrun/taskrun.go, and the PipelineRun reconciler code is found in pkg/reconciler/pipelinerun/pipelinerun.go.
+
+### Webhooks
+
+All Tekton CRDs use validating [admission webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/)
+to validate instances of CRDs. Some CRDs also use mutating admission webhooks to set default values for some fields.
```

Comment:
would be useful to have a link to a mutating webhook configuration

Reply:
are you referring to this? https://github.com/tektoncd/pipeline/blob/main/config/500-webhooks.yaml#L48-L65

Reply:
I meant some docs that describe what a mutating webhook is. https://banzaicloud.com/blog/k8s-admission-webhooks/ was a good read

Reply:
done